models/articles.model.js

```diff
@@ -1,16 +1,73 @@
 const db = require("../db/connection");
 
-exports.getArticlesModel = () => {
+exports.getArticlesModel = (sort_by = "date", order = "desc", topic) => {
   let SQL = `SELECT articles.author, title, articles.article_id, 
   topic, articles.created_at, articles.votes, 
   COUNT(comments.article_id):: INTEGER AS comment_count
   FROM articles
-  LEFT JOIN comments ON articles.article_id = comments.article_id
-  GROUP BY articles.article_id
-  ORDER BY articles.created_at DESC
-  ;
+  LEFT JOIN comments ON articles.article_id = comments.article_id 
   `;
-  return db.query(SQL).then(({ rows }) => rows);
+
+  const queryValues = [];
+
+  const validSortBy = [
+    "article_id",
+    "author",
+    "title",
+    "topic",
+    "date",
+    "votes",
+    "comment_count",
+  ];
+  const validOrder = ["asc", "desc"];
+
+  if (!validSortBy.includes(sort_by) || !validOrder.includes(order)) {
+    return Promise.reject({ status: 400, msg: "Bad Request" });
+  }
+
+  if (topic !== undefined) {
+    SQL += ` WHERE topic = $1 GROUP BY articles.article_id`;
+    queryValues.push(topic);
+  } else {
+    SQL += ` GROUP BY articles.article_id`;
+  }
+
+  if (sort_by === "article_id") {
+    SQL += ` ORDER BY articles.article_id`;
+  } else if (sort_by === "author") {
+    SQL += ` ORDER BY articles.author`;
+  } else if (sort_by === "title") {
+    SQL += ` ORDER BY title`;
+  } else if (sort_by === "topic") {
+    SQL += ` ORDER BY topic`;
+  } else if (sort_by === "comment_count") {
+    SQL += ` ORDER BY comment_count`;
+  } else if (sort_by === "votes") {
+    SQL += ` ORDER BY votes`;
+  } else {
+    SQL += ` ORDER BY articles.created_at`;
+  }
+
+  if (order === "asc") {
+    SQL += ` ASC;`;
+  } else {
+    SQL += ` DESC;`;
+  }
+
+  return db.query(SQL, queryValues).then(({ rows }) => rows);
+};
+
+exports.checkArticleQueryExists = (reqQuery) => {
+  const queryArray = Object.keys(reqQuery);
+  const validQuery = ["sort_by", "topic", "order"];
+  return new Promise((resolve, reject) => {
+    if (queryArray.length === 0) return resolve(true);
+    queryArray.forEach((q) => {
+      if (!validQuery.includes(q)) {
+        return reject({ status: 400, msg: "Bad Request" });
```

Comment:
While I don't think this is 100% necessary it is very good that you're thinking about how to throw errors in these scenarios - it's very impressive!

However I will say that `q` is not a semantic variable name and should be renamed to `query` or something